File: brahe/orbit_dynamics/relativity.py
"""This orbit dynamics submoduble provides functions for computing acceleration 
due to special relativity
"""

import numba
import numpy as np

import brahe.constants as _const

@numba.jit(nopython=True, cache=True)
def accel_relativity(x:np.ndarray) -> np.ndarray:
    '''Computes perturbation accleration of a satellite in the Inertial frame
    due to the combined effects of special and general relativity.

    Args:
        x (:obj:`np.ndarray`): Satellite Cartesean state in the inertial reference frame [m; m/s]

    Returns:
        np.ndarray: Satellite acceleration due to relativity. [m/s^2]

    References:
        1. O. Montenbruck, and E. Gill, _Satellite Orbits: Models, Methods and Applications_, 2012, p.110-112.
    '''
    # Extract state variables
    r = x[0:3]
    v = x[3:6]

    # Intermediate computations
    norm_r = np.linalg.norm(r)
    r2     = norm_r**2

    norm_v = np.linalg.norm(v)
    v2     = norm_v**2

    c  = _const.C_LIGHT
    c2 = c**2

    # Compute unit vectors
    er = r/norm_r
    ev = v/norm_v

    # Compute perturbation and return
    a_rel = _const.GM_EARTH/r2 * ( (4*_const.GM_EARTH/(c2*norm_r) - v2/c2)*er + 4*v2/c2*np.dot(er, ev)*ev)

    return a_rel

File: src/orbit_dynamics/relativity.rs
/*!
 Acceleration due to non-Newtionian gravitational effects. Specifically, special and general
relativity.
 */

use nalgebra::{Vector3, Vector6};

use crate::constants::{C_LIGHT, GM_EARTH};

/// Calculate the acceleration due to special and general relativity for an Earth orbiting object.
///
/// # Arguments
///
/// - `x_object`: State vector of the object in the ECI frame.
///
/// # Returns
///
/// - `a_relativity` : Acceleration due to special and general relativity.
///
/// # Examples
///
/// ```
/// use brahe::coordinates::state_osculating_to_cartesian;
/// use brahe::orbit_dynamics::acceleration_relativity;
/// use nalgebra::Vector6;
/// use brahe::R_EARTH;
///
/// let x_object = Vector6::new(R_EARTH + 500.0e3, 0.0, 0.0, 0.0, 0.0, 0.0);
/// let a_relativity = acceleration_relativity(x_object);
/// ```
pub fn acceleration_relativity(x_object: Vector6<f64>) -> Vector3<f64> {
    // Extract state variables
    let r = x_object.fixed_rows::<3>(0);
    let v = x_object.fixed_rows::<3>(3);

    // Intermediate computations
    let norm_r = r.norm();
    let r2 = norm_r.powi(2);
    let norm_v = v.norm();
    let v2 = norm_v.powi(2);
    let c2 = C_LIGHT.powi(2);

    // Compute unit vectors
    let er = r / norm_r;
    let ev = v / norm_v;

    // Compute perturbation acceleration and return
    GM_EARTH / r2 * ((4.0 * GM_EARTH / (c2 * norm_r) - v2 / c2) * er + 4.0 * v2 / c2 * er.dot(&ev) * ev)
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_acceleration_relativity() {
        // TODO: Add tests
    }
}